odxtools/odxlink.py

```diff
@@ -0,0 +1,223 @@
+
+from typing import Dict, Union, Optional, Any
+from dataclasses import dataclass, field, replace
+from xml.etree.ElementTree import Element
+
+@dataclass(frozen=True)
+class OdxLinkRef:
+    """A reference to an ODX object.
+
+    OdxLinkRefs are similar to OdxLinkIds, with the difference that the document
+    of the referred-to object does not need to be specified. (In this case, it
+    means "use the document where the referring object is located in")
+    """
+
+    #: The local identifier of the object which is referred to
+    ref_id : str
+
+    #: The document fragments to which the `ref_id` refers to (in reverse order)
+    ref_docs : List[OdxDocFragment]
+
+    @staticmethod
+    def from_et(et: Optional[Element], source_doc_frags: List[OdxDocFragment]) -> Optional["OdxLinkRef"]:
+        """Construct an OdxLinkRef for a given XML node (ElementTree object).
+
+        Returns None if the given XML node does not represent a reference.
+        """
+
+        if et is None:
+            return None
+
+        id_ref = et.attrib.get("ID-REF")
+        if id_ref is None:
+            return None
+
+        doc_ref = et.attrib.get("DOCREF")
+        doc_type = et.attrib.get("DOCTYPE")
+
+        assert \
+            (doc_ref is not None and doc_type is not None) or \
+            (doc_ref is None and doc_type is None), \
+            "DOCREF and DOCTYPE must both either be specified or omitted"
+
+        # if the target document fragment is specified by the
+        # reference, use it, else use the document fragment containing
+        # the reference.
+        if doc_ref is not None:
+            doc_frags = [OdxDocFragment(doc_ref, doc_type)]
+        else:
+            doc_frags = source_doc_frags
+
+        return OdxLinkRef(id_ref, doc_frags)
+
+    @staticmethod
+    def from_id(odxid: OdxLinkId) -> "OdxLinkRef":
+        """Construct an OdxLinkRef for a given OdxLinkId.
+        """
+        return OdxLinkRef(odxid.local_id, odxid.doc_fragments)
+
+    def __contains__(self, id: OdxLinkId) -> bool:
+        """
+        Returns true iff a given OdxLinkId object is referenced.
+        """
+
+        # we must reference at to at least of the ID's document
+        # fragments
+        if not any([ref_doc in id.doc_fragments for ref_doc in self.ref_docs]):
+            return False
+
+        # the local ID of the reference and the object ID must match
+        return id.local_id == self.ref_id
+
+class OdxLinkDatabase:
+    """
+    A database holding all objects which ehibit OdxLinkIds
+
+    This can resolve references to such.
+    """
+
+    def __init__(self):
+        self._db: Dict[OdxDocFragment, Dict[OdxLinkId, Any]] = {}
+
+    def resolve(self, ref: OdxLinkRef) -> Any:
+        """
+        Resolve a reference to an object
+
+        If the database does not contain any object which is referred to, a
+        KeyError exception is raised.
+        """
+        assert isinstance(ref, OdxLinkRef)
```

Comment:
if res is None return None?

Reply:
this is intentional: it makes implementation errors much more obvious (and if this happens, some calling code needs to be fixed...)